pkg/resource/predicates.go

```diff
@@ -146,3 +146,17 @@ func IsNamed(name string) PredicateFn {
 		return mo.GetName() == name
 	}
 }
+
+// DesiredStateChanged accepts objects that have changed their desired state, i.e.
+// the state that is not managed by the controller.
+// To be more specific, it accepts update events that have changes in one of the followings:
+// - `metadata.annotations`
+// - `metadata.labels`
+// - `spec`
+func DesiredStateChanged() predicate.Predicate {
```

Comment:
I notice almost everything else in this package returns our `PredicateFn` type, whereas this returns an upstream `predicate.Predicate` interface. It seems like all three of the `predicate.Or`-ed predicates only implement `Update`.

I'm wondering whether it makes sense for this to return a `PredicateFn` for consistency/compatibility with the rest of this package.

Alternatively, I think a lot of what this package does has been implemented upstream since we wrote it. For example `predicate.AnyOf` seems about the same as `predicate.Or`. Perhaps we should be deprecating much of this package in favor of upstream functionality?

Reply:
> I'm wondering whether it makes sense for this to return a PredicateFn for consistency/compatibility with the rest of this package.

This was something I've considered in the beginning; however, it looks like this is not possible since the predicates we are adding need both [the old object and new object](https://github.com/kubernetes-sigs/controller-runtime/blob/97acd6444ff1ec87caf7d2077a9056ea56b6ba0e/pkg/predicate/predicate.go#L171) (e.g. to see if generation changed) but the `PredicateFn` only receives a single object (i.e. `ObjectNew`).

> Alternatively, I think a lot of what this package does has been implemented upstream since we wrote it. For example predicate.AnyOf seems about the same as predicate.Or. Perhaps we should be deprecating much of this package in favor of upstream functionality?

I see some potential value in having this specific predicate in the runtime:

- Instead of expecting each provider to combine the three in the same way, they can rely on what runtime has.
- We may want to change things under the hood on how we define the "desired state" for a Crossplane MR. A good example of this is one of the proposed solutions [here](https://github.com/crossplane/crossplane/issues/3840#issuecomment-1461758879). If we decide that some specific annotation should not be part of the desired state, we can add exceptions in this central place.

Regarding the deprecation of the rest, I wanted to see how they are used today and searched them one by one using a query as below and couldn't find any usage for any of them: 

https://github.com/search?q=org%3Acrossplane+OR+org%3Aupbound+OR+org%3Acrossplane-contrib+IsManagedKind&type=code

Am I missing something, or are they not used at all already?

Reply:
> I see some potential value in having this specific predicate in the runtime

I agree - to be clear I was only suggesting removing (most of?) the _other_ stuff in this package. Not the new predicate.

I'm a bit surprised the other stuff aren't used, but I trust your search. Maybe they haven't been used since the old resource claim reconcilers or something? Given they aren't used, would you mind adding `// Deprecated` comments to them in a separate commit? Then we can go through and remove a bunch of deprecated things next release.

Reply:
And I trust Github's search :) 

Done. 
Some have a single usage, like [IsNamed](https://github.com/upbound/universal-crossplane/blob/7403c2fe233e5335fe1251a031df36399abd1d1e/internal/controllers/billing/setup.go#L54), and I still added deprecation notice since it can live in the relevant codebase, no need to have shared.